File: cpp_example/object_invariant.cpp
#include<cassert>

class ValueStock
{
public:
	ValueStock(float price, float book_value) : 
		mprice(price),
		mbook_value(book_value)
	{}

	void buy(){
		this->invariant();
		...
		this->invariant();
	}

	void sell(){
		this->invariant();
		...
		this->invariant();
	}

private:
	void invariant(){
		assert(this->mprice < this->mbook_value); //Value stocks must have prices below book values
	}
};

	

File: python_example/object_invariant.py
class ValueStock(object):
	"""A stock is a value stock if its Price to Book ratio is below 1"""
	def __init__(self, price, book_value):
		self._price = price
		self._book_value = book_value
		self._invariant()

	def _invariant_(self):
		assert self._price < self._book_value, "Value stocks must have prices below book values"

	def buy(self):
		self._invariant()
		...
		self._invariant()

	def sell(self):
		self._invariant()
		...
		self._invariant()
